## Linear time Closed item set Miner
LCM looks for closed itemset with respect to an input minimum support

#### load the chess dataset

In [1]:
from skmine.datasets.fimi import fetch_chess
chess = fetch_chess()
chess.head()

0    [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25...
1    [1, 3, 5, 7, 9, 12, 13, 15, 17, 19, 21, 23, 25...
2    [1, 3, 5, 7, 9, 12, 13, 16, 17, 19, 21, 23, 25...
3    [1, 3, 5, 7, 9, 11, 13, 15, 17, 20, 21, 23, 25...
4    [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25...
Name: chess, dtype: object

In [2]:
chess.shape

(3196,)

#### fit_discover()
fit_discover makes pattern discovery more user friendly by outputting pretty formatted
patterns, instead of the traditional tabular format used in the `scikit` community

In [3]:
from skmine.itemsets import LCM
lcm = LCM(min_supp=2000, n_jobs=4)
# minimum support of 2000, running on 4 processes
%time patterns = lcm.fit_discover(chess)

CPU times: user 56.3 ms, sys: 24.7 ms, total: 81 ms
Wall time: 1.47 s


In [4]:
patterns.shape

(68967, 2)

This format in which patterns are rendered makes post hoc analysis easier

Here we filter patterns with a length strictly superior to 3

In [5]:
patterns[patterns.itemset.map(len) > 3]

itemset support
14         [29, 40, 52, 58]    3143
22         [29, 52, 58, 60]    3124
26         [40, 52, 58, 60]    3112
28         [29, 40, 58, 60]    3110
29         [29, 40, 52, 60]    3100
...                     ...     ...
68960      [15, 52, 58, 60]    2003
68962      [15, 29, 58, 60]    2002
68964      [29, 40, 58, 70]    2006
68965      [29, 40, 52, 70]    2001
68966  [29, 40, 52, 58, 70]    2000

[67716 rows x 2 columns]

`Note`

Even when setting a very high minimum support threshold, we discovered more than 60K from only 3196 original transactions.
This is a good illustration of the so-called **pattern explosion problem**

------------
We could also get the top-k patterns in terms of supports, with a single line of code

In [6]:
patterns.nlargest(10, columns=['support'])  # top 10 patterns

TypeError: Column 'support' has dtype object, cannot use method 'nlargest' with this dtype